File: src/Devices/MbusDevice.fs
namespace Mbus

open System.IO
open System.Threading
open System.Collections.Generic
open Mbus.Devices
open Mbus.Records

type MbusDevice(stream: Stream, address: byte) =
    let mutable _dataRecords : Record list = []
    let _lock = obj()

    member this.SetRecords(records: IEnumerable<Record>) =
        let recordsList = List.ofSeq records
        lock _lock (fun () ->
            _dataRecords <- recordsList
        )

    member this.RunAsync(ct: CancellationToken) =
        let initialState = DeviceLogic.initialState address
        let getData () = lock _lock (fun () -> _dataRecords)
        DeviceRuntime.runAsync stream getData initialState ct

File: src/MbusDevice/MbusDevice.cs
using Mbus.Devices;
using Mbus.Records;
using Mbus.UserData;
using Microsoft.FSharp.Collections;
using Microsoft.FSharp.Core;

namespace Mbus;

public class MbusDevice
{
    private readonly Lock _lock = new();
    private readonly List<DataRecord> _userData = [];

    public void UpdateUserData(Action<IUserDataBuilder> updateAction)
    {
        lock (_lock)
        {
            _userData.Clear();
            var builder = new UserDataBuilder(_userData);
            updateAction(builder);
        }
    }

    public Task RunAsync(
        Stream stream,
        int idNumber,
        string mfrCode,
        MbusDeviceType deviceType,
        byte version,
        byte primaryAddress,
        CancellationToken ct)
    {
        var address = MbusAddress.Create(idNumber, mfrCode, version, deviceType);

        if (address.IsError)
            throw new ArgumentException(address.ErrorValue);

        var initialState = DeviceLogic.initialState(primaryAddress, address.ResultValue);

        return DeviceRuntime.runAsync(stream, GetDataFunc, initialState, ct);
    }

    private FSharpFunc<Unit, FSharpList<Record>> GetDataFunc =>
        FuncConvert.FromFunc(() =>
        {
            lock (_lock)
            {
                return ListModule.OfSeq(_userData.Select(Record.NewData));
            }
        });
}